casanovo/denovo/model.py

```diff
@@ -989,6 +989,190 @@ def configure_optimizers(
         return [optimizer], {"scheduler": lr_scheduler, "interval": "step"}
 
 
+class DBSpec2Pep(Spec2Pep):
+    """
+    Inherits Spec2Pep
+
+    Hijacks teacher-forcing implemented in Spec2Pep and
+    uses it to predict scores between a spectra and associated peptide.
+    Input format is .mgf, with comma-separated targets
+    and decoys in the SEQ field. Decoys should have a prefix of "decoy_".
+    """
+
+    num_pairs = None  # Modified to be predict_batch_size from config
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def predict_step(self, batch, *args):
+        batch_res = []
+        for (
+            indexes,
+            t_or_d,
+            peptides,
+            precursors,
+            encoded_ms,
+        ) in self.smart_batch_gen(batch):
+            pred, truth = self.decoder(peptides, precursors, *encoded_ms)
+            sm = torch.nn.Softmax(dim=2)
+            pred = sm(pred)
+            score_result, per_aa_score = _calc_match_score(
+                pred, truth
+            )  # Calculate the score between spectra + peptide list
+            batch_res.append(
+                (
+                    indexes,
+                    t_or_d,
+                    peptides,
+                    score_result.cpu().detach().numpy(),
+                    per_aa_score.cpu().detach().numpy(),
+                    precursors.cpu().detach().numpy(),
+                )
+            )
+        return batch_res
+
+    def smart_batch_gen(self, batch):
```

Comment:
This was the best way I could think of to prevent unnecessary forward passes of the encoder. It takes in a datapoint just like _de novo_ Casanovo would, encodes the spectrum once, and then replicates it once for each candidate. This is then considered one "block" of PSMs. However, this is usually not enough to saturate GPU memory, so I repeat this process until a certain number of PSMs are put together.

This makes it so that a forward pass encodes the spectrum just once per spectrum, instead of once per PSM.